Expose connection hot zones in `GraphNode`

mechPenSketch opened the pull request:
<!--
Pull requests should always be made for the `master` branch first, as that's
where development happens and the source of all future stable release branches.

Relevant fixes are cherry-picked for stable branches as needed.

Do not create a pull request for stable branches unless the change is already
available in the `master` branch and it cannot be easily cherry-picked.
Alternatively, if the change is only relevant for that branch (e.g. rendering
fixes for the 3.2 branch).
-->
Following [my proposal](https://github.com/godotengine/godot-proposals/issues/3009), I thought I could make a small edit to godot's source code to expose GraphNodes' hot zones.

Currently, a hot zone is defined as a rectangular area where its middle is the port's position. The hot zone's width is twice the GraphEdit's theme property `port_grab_distance_horizontal`, and its height twice the `port_grab_distance_vertical`.

With this pull, the user can override what the hot zone is with `is_in_input_hotzone(graph_node, child_index, mouse_position)` and `is_in_output_hotzone(graph_node, child_index, mouse_position)`. For my case, the hot zones are the leftmost quarter of the GraphNode for the output and the rightmost quarter for the input.

Although I don't know anyone else who needs this, this could potentially be useful to those who are designing with GraphEdits and GraphNodes, and their styles could call for a unique shape for their hot zones.

Hopefully, this will take effect by the next version so that I can continue on my FSM plugin.

mechPenSketch committed:
Split Parts Checking For Hot Zones Into New Functions.

mechPenSketch committed:
Update _bind_methods()

This is to expose the newly-added functions.

mechPenSketch committed:
Documenting Changes

Added two new methods.

pycbouh commented on scene/gui/graph_edit.h:
```diff
@@ -184,7 +184,9 @@ class GraphEdit : public Control {
 
-	bool is_in_hot_zone(const Vector2 &pos, const Vector2 &p_mouse_pos);
+	bool is_in_input_hotzone(GraphNode gn, int i, const Vector2 &p_mouse_pos)
+	bool is_in_output_hotzone(GraphNode gn, int i, const Vector2 &p_mouse_pos)
```
Those two lines are missing semicolons, hence the build failing on every platform. Make sure you test your changes locally, that they compile and that they work and don't cause obvious side-effects.

pycbouh commented on scene/gui/graph_edit.h:
```diff
@@ -184,7 +184,9 @@ class GraphEdit : public Control {
 	void _top_layer_input(const Ref<InputEvent> &p_ev);
 
-	bool is_in_hot_zone(const Vector2 &pos, const Vector2 &p_mouse_pos);
+	bool is_in_input_hotzone(GraphNode gn, int i, const Vector2 &p_mouse_pos)
```
The convention is to use a `p_` prefix for parameter names (see `p_mouse_pos`). Sometimes the original code is not following this convention, but if you touch it, it's worth making sure you do.

Also `gn` and `i` are not a great parameter name in general, you need to be more verbose.

And `GraphNode` being a node has to be passed by pointer, so `GraphNode *gn` (and this would likely require some changes to the method implementation as well).

pycbouh commented on scene/gui/graph_edit.cpp:
```diff
@@ -1704,6 +1709,8 @@ void GraphEdit::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("_gui_input"), &GraphEdit::_gui_input);
 	ClassDB::bind_method(D_METHOD("_update_scroll_offset"), &GraphEdit::_update_scroll_offset);
+	ClassDB::bind_method(D_METHOD("is_in_input_hotzone", "graph_node", "child_index", "mouse_position"), &GraphEdit::is_in_input_hotzone);
+	ClassDB::bind_method(D_METHOD("is_in_output_hotzone", "graph_node", "child_index", "mouse_position"), &GraphEdit::is_in_output_hotzone);
```
I don't think `child_index` is an appropriate name. Looking at where it comes from, `connection_index`, maybe?

mechPenSketch replied:
I thought `child index` was appropriate as I though it takes a child of a GraphNode for a slot to appear. For now, I've changed it to `slot_index` as I think it fits more.

YuriSizov reviewed (changes requested):
I think we could expose something along those lines. You still have issues with code and code style which need to be fixed.

I would also amend the documentation, but before that I believe some more code changes need to be made that would affect the method descriptions. Correct me if I'm wrong, but I think you meant to expose some sort of virtual method that can be overridden? Virtual methods that can be overridden in the engine need to be marked with the `virtual` keyword. Virtual methods that we expose to the scripting API are bound to the ClassDB in a special way, and we usually prefix them with `_`. Check out how `_set_create_options` is implemented in https://github.com/godotengine/godot/blob/master/editor/editor_resource_picker.cpp.

And don't forget to squash your commits and only leave one with a descriptive message (your PR name is probably fine for that).

pycbouh commented:
And to answer your questions from the proposal:

> What's with the ands and asterisks in the parameters?

Ampersands designate parameters passed by references. Asterisks designate parameters passed by pointer. There are intricacies where one would use which, so if you're not sure just follow the convention you see in existing code.

> And should I set one of the parameters as a constant if I'm going to expose the function?

What you ended up doing seems to be fine, but not all parameters need to be consts in general.

mechPenSketch committed:
Fixed Missing Semi-Colon, Renamed Parameters

mechPenSketch committed:
Corrected Parameter Names

This is to follow the conventional format.

mechPenSketch commented:
> Correct me if I'm wrong, but I think you meant to expose some sort of virtual method that can be overridden? Virtual methods that can be overridden in the engine need to be marked with the `virtual` keyword. Virtual methods that we expose to the scripting API are bound to the ClassDB in a special way, and we usually prefix them with `_`. Check out how `_set_create_options` is implemented in https://github.com/godotengine/godot/blob/master/editor/editor_resource_picker.cpp.

I don't intend `is_in_input_hotzone(...)` and `is_in_output_hotzone(...)` to be virtual methods. Even if they are to be overridden, it's still possible that there will be users who wants to use the original hot zones.

pycbouh commented:
> Even if they are to be overridden, it's still possible that there will be users who wants to use the original hot zones.

In that case you can expose both the normal method that falls back to the default implementation, and the virtual API method that is to be implemented by the user and if so will be called by the normal method instead of the default implementation. Your main purpose was to provide a way to override the default behavior, if I understand your proposal correctly. This PR needs to address that first, and not a hypothetical need to use the default implementation. So making it clear that there is a possibility for the override in the API is important, and the normal method is secondary.

PS. You still have many issues with code that make me assume you don’t build it locally. This will take a very long time if you don’t test it (and frankly we cannot fully trust a PR that was not tested by its author).

mechPenSketch committed:
Fixed Missing Variable

There were later lines that require Vector2 pos that was previously removed.

mechPenSketch commented:
I started out using Notepad++ thinking I could just tweak a few codes, but I switched to IntelliJ now that the change is getting more complex. Is Build > Rebuild Project the correct way of building locally?

pycbouh commented:
> Is Build > Rebuild Project the correct way of building locally?

Depending on your operating system and the IDE you use there are different steps involved into setting up your working environment, but when you do it would be as simple as that, yes.

Here's a set of articles on compiling the engine and setting up the build tools:
https://docs.godotengine.org/en/stable/development/compiling/index.html

Here's a set of guides on setting up some of the more popular IDEs and code editors for C++ development:
https://docs.godotengine.org/en/stable/development/cpp/configuring_an_ide/index.html

fire commented:
I think the feature is a good proposal.

Need to pass all the tests though to be merged.

mechPenSketch commented:
^ Thanks for your words of encouragement. Unfortunately, I'm currently stuck building Godot locally, and I don't think it's error in my code this time.

This is the output when I type in `scons platform=windows vsproj=yes` on Godot's root folder:
>scons: Reading SConscript files ...
Configuring for Windows: target=debug, bits=default
Found MSVC version 14.2, arch amd64, bits=64
Note: Building a debug binary (which will run slowly). Use `target=release_debug` to build an optimized release binary.
YASM is necessary for WebM SIMD optimizations.
WebM SIMD optimizations are disabled. Check if your CPU architecture, CPU bits or platform are supported!
Checking for C header file mntent.h... (cached) no
scons: done reading SConscript files.
scons: Building targets ...
[Initial build] Compiling ==> thirdparty\vulkan\loader\loader.c
loader.c
thirdparty\vulkan\loader\loader.c(6059): error C2065: 'IDXGIFactory6': undeclared identifier
thirdparty\vulkan\loader\loader.c(6059): error C2065: 'dxgi_factory': undeclared identifier
thirdparty\vulkan\loader\loader.c(6059): error C2106: '=': left operand must be l-value
thirdparty\vulkan\loader\loader.c(6060): error C2065: 'IID_IDXGIFactory6': undeclared identifier
thirdparty\vulkan\loader\loader.c(6060): error C2065: 'dxgi_factory': undeclared identifier
thirdparty\vulkan\loader\loader.c(6063): error C2065: 'dxgi_factory': undeclared identifier
thirdparty\vulkan\loader\loader.c(6063): error C2223: left of '->lpVtbl' must point to struct/union
thirdparty\vulkan\loader\loader.c(7088): error C2065: 'IDXGIFactory6': undeclared identifier
thirdparty\vulkan\loader\loader.c(7088): error C2065: 'dxgi_factory': undeclared identifier
thirdparty\vulkan\loader\loader.c(7088): error C2106: '=': left operand must be l-value
thirdparty\vulkan\loader\loader.c(7089): error C2065: 'IID_IDXGIFactory6': undeclared identifier
thirdparty\vulkan\loader\loader.c(7089): error C2065: 'dxgi_factory': undeclared identifier
thirdparty\vulkan\loader\loader.c(7106): error C2065: 'dxgi_factory': undeclared identifier
thirdparty\vulkan\loader\loader.c(7106): error C2223: left of '->lpVtbl' must point to struct/union
thirdparty\vulkan\loader\loader.c(7106): error C2065: 'DXGI_GPU_PREFERENCE_UNSPECIFIED': undeclared identifier
thirdparty\vulkan\loader\loader.c(7190): error C2065: 'dxgi_factory': undeclared identifier
thirdparty\vulkan\loader\loader.c(7190): error C2223: left of '->lpVtbl' must point to struct/union
scons: *** [thirdparty\vulkan\loader\loader.windows.tools.64.obj] Error 2
scons: building terminated because of errors.
[Time elapsed: 00:00:08.496]

akien-mga commented:
Your MinGW version is too old and doesn't include the necessary DirectX headers used by the Vulkan loader. You should update it to the latest MinGW version (at least 8.0.0).
Note that the MinGW version is not the same as the GCC version provided with MinGW.

mechPenSketch commented:
^ My Visual Studio Community 2019 is already up-to-date, and Windows 10 SDK. However, the Windows 10 SDK is the earliest among the five versions available. Could it be that I've installed the wrong versions?

akien-mga commented:
> However, the Windows 10 SDK is the earliest among the five versions available. Could it be that I've installed the wrong version?

mechPenSketch commented:
Ok, now that the building issue is out of the way, I can deal with errors by my own code.

I don't know what to make of this when the building comes to `graphic_edit.cpp`:
>C:\Users\User\Documents\Godot\Raw\godot\core/variant/binder_common.h(49): error C2440: 'return': cannot convert from 'const Variant' to 'T'
        with
        [
            T=GraphNode *
        ]
C:\Users\User\Documents\Godot\Raw\godot\core/variant/binder_common.h(49): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
C:\Users\User\Documents\Godot\Raw\godot\core/variant/binder_common.h(48): note: while compiling class template member function 'T VariantCaster<T>::cast(const Variant &)'
        with
        [
            T=GraphNode *
        ]
C:\Users\User\Documents\Godot\Raw\godot\core/variant/binder_common.h(177): note: see reference to function template instantiation 'T VariantCaster<T>::cast(const Variant &)' being compiled
        with
        [
            T=GraphNode *
        ]
C:\Users\User\Documents\Godot\Raw\godot\core/variant/binder_common.h(177): note: see reference to class template instantiation 'VariantCaster<T>' being compiled
        with
        [
            T=GraphNode *
        ]
C:\Users\User\Documents\Godot\Raw\godot\core/variant/binder_common.h(168): note: while compiling class template member function 'T VariantCasterAndValidate<T>::cast(const Variant **,uint32_t,Callable::CallError &)'
        with
        [
            T=GraphNode *
        ]
C:\Users\User\Documents\Godot\Raw\godot\core/variant/binder_common.h(644): note: see reference to function template instantiation 'T VariantCasterAndValidate<T>::cast(const Variant **,uint32_t,Callable::CallError &)' being compiled
        with
        [
            T=GraphNode *
        ]
C:\Users\User\Documents\Godot\Raw\godot\core/variant/binder_common.h(640): note: see reference to class template instantiation 'VariantCasterAndValidate<GraphNode *>' being compiled
C:\Users\User\Documents\Godot\Raw\godot\core/variant/binder_common.h(420): note: see reference to function template instantiation 'void call_with_variant_args_ret_helper<T,R,GraphNode*,int,const Vector2&,0,1,2>(T *,R (__cdecl GraphEdit::* )(GraphNode *,int,const Vector2 &),const Variant **,Variant &,Callable::CallError &,IndexSequence<0,1,2>)' being compiled
        with
        [
            T=GraphEdit,
            R=bool
        ]
C:\Users\User\Documents\Godot\Raw\godot\core/object/method_bind.h(449): note: see reference to function template instantiation 'void call_with_variant_args_ret_dv<T,R,GraphNode*,int,const Vector2&>(T *,R (__cdecl GraphEdit::* )(GraphNode *,int,const Vector2 &),const Variant **,int,Variant &,Callable::CallError &,const Vector<Variant> &)' being compiled
        with
        [
            T=GraphEdit,
            R=bool
        ]
C:\Users\User\Documents\Godot\Raw\godot\core/object/method_bind.h(449): note: while compiling class template member function 'Variant MethodBindTR<T,R,GraphNode *,int,const Vector2 &>::call(Object *,const Variant **,int,Callable::CallError &)'
        with
        [
            T=GraphEdit,
            R=bool
        ]
C:\Users\User\Documents\Godot\Raw\godot\core/object/method_bind.h(478): note: see reference to class template instantiation 'MethodBindTR<T,R,GraphNode *,int,const Vector2 &>' being compiled
        with
        [
            T=GraphEdit,
            R=bool
        ]
C:\Users\User\Documents\Godot\Raw\godot\core/object/class_db.h(244): note: see reference to function template instantiation 'MethodBind *create_method_bind<GraphEdit,bool,GraphNode*,int,const Vector2&>(R (__cdecl GraphEdit::* )(GraphNode *,int,const Vector2 &))' being compiled
        with
        [
            R=bool
        ]
scene\gui\graph_edit.cpp(1715): note: see reference to function template instantiation 'MethodBind *ClassDB::bind_method<MethodDefinition,bool(__cdecl GraphEdit::* )(GraphNode *,int,const Vector2 &)>(N,M)' being compiled
        with
        [
            N=MethodDefinition,
            M=bool (__cdecl GraphEdit::* )(GraphNode *,int,const Vector2 &)
        ]
scons: *** [scene\gui\graph_edit.windows.tools.64.obj] Error 2
scons: building terminated because of errors.
[Time elapsed: 00:08:27.356]

Could it be that GraphNode can't be accepted as a method input (function parameter)? Maybe I can cross-refer with other methods to see whether there are other ways to get the GraphNode from other property type...

akien-mga commented:
> Could it be that GraphNode can't be accepted as a method input (function parameter)? Maybe I can cross-refer with other methods to see whether there are other ways to get the GraphNode from other property type...

Yes, that's a design limitation of the binding system. The only node pointer types available for bindings are `Node *` and `Control *`.

But what's more common would be to use a `Ref<GraphNode>` reference instead, which should work fine for the binding if that's suitable for your use case.

mechPenSketch commented:
^ Do you have an example where refences can be used for Node and Control's extensions? From reading `graphic_edit.cpp` alone, they seem to be used for those extended from Resource.

akien-mga commented:
> ^ Do you have an example where refences can be used for Node and Control's extensions? From reading `graphic_edit.cpp` alone, they seem to be used for those extended from Resource.

You're right, my bad. `Ref<>` is for Resource types. So you should use `Control *` here, and cast to `GraphNode *` in the relevant method to validate the input (`GraphNode *gnode = Object::cast_to<GraphNode>(p_node);`). It's `nullptr` if the object is not a `GraphNode`.

mechPenSketch committed:
Made the New Methods Virtual

mechPenSketch commented:
Since I'm advised to convert `is_in_input_hotzone(...)` and its output counterpart into virtual methods, I might as well do that now along the way. I originally want to correct the error I'm currently at first before continuing, but I think I can make the jump.

mechPenSketch committed:
Minor Format Fix

mechPenSketch committed:
Further Documentation Changes

This is in following to converting is_in_input_hotzone(...) and its output counterpart to virtual methods.

mechPenSketch commented:
@pycbouh has something to say about my documentation. While I have updated it to account for the new methods being virtual, there might be other amendments I didn't account for, plus I have one workflow failure in my repo.

The failure seems to be from `_is_in_input_hotzone` and output counterpart being set as booleans:
```
Run echo "Running --doctool to see if this changes the public API without updating the documentation."
Running --doctool to see if this changes the public API without updating the documentation.
If a diff is shown, it means that your code/doc changes are incomplete and you should update the class reference with --doctool.


diff --git a/doc/classes/GraphEdit.xml b/doc/classes/GraphEdit.xml
index 99e0e15..225705d 100644
--- a/doc/classes/GraphEdit.xml
+++ b/doc/classes/GraphEdit.xml
@@ -11,7 +11,7 @@
 	</tutorials>
 	<methods>
 		<method name="_is_in_input_hotzone" qualifiers="virtual">
-			<return type="bool" />
+			<return type="void" />
 			<argument index="0" name="graph_node" type="NodePath" />
 			<argument index="1" name="slot_index" type="int" />
 			<argument index="2" name="mouse_position" type="Vector2" />
@@ -21,7 +21,7 @@
 			</description>
 		</method>
 		<method name="_is_in_output_hotzone" qualifiers="virtual">
-			<return type="bool" />
+			<return type="void" />
 			<argument index="0" name="graph_node" type="NodePath" />
 			<argument index="1" name="slot_index" type="int" />
 			<argument index="2" name="mouse_position" type="Vector2" />
Error: Process completed with exit code 1.
```
I'm cross-referring with other nodes to see how virtual methods are added when it comes to different datatype.

mechPenSketch committed:
Change Datatypes of the New Methods to Bool

mechPenSketch committed:
Fixed Parameters in Adding Virtual Methods

mechPenSketch commented:
I've fixed `add_virtual_method(...)`'s parameters and the Linux build succeed. However, now it's the iOS build that fails. Doesn't seem that it's the code issue, though. Not sure how to resolve it.

KoBeWi commented on scene/gui/graph_edit.cpp:
```diff
@@ -1704,6 +1722,8 @@ void GraphEdit::_bind_methods() {
 
 	ClassDB::bind_method(D_METHOD("_gui_input"), &GraphEdit::_gui_input);
 	ClassDB::bind_method(D_METHOD("_update_scroll_offset"), &GraphEdit::_update_scroll_offset);
+	ClassDB::add_virtual_method(get_class_static(), MethodInfo(Variant::BOOL, "_is_in_input_hotzone", PropertyInfo(Variant::NODE_PATH, "graph_node"), PropertyInfo(Variant::INT, "slot_index"), PropertyInfo(Variant::VECTOR2, "mouse_position")));
+	ClassDB::add_virtual_method(get_class_static(), MethodInfo(Variant::BOOL, "_is_in_output_hotzone", PropertyInfo(Variant::NODE_PATH, "graph_node"), PropertyInfo(Variant::INT, "slot_index"), PropertyInfo(Variant::VECTOR2, "mouse_position")));
```
You could just use `BIND_VMETHOD()`

mechPenSketch replied:
True, but this is the solution I've already found before I decided to replace `add_virtual_method(...)` with `BIND_VMETHOD()(...)`. As long as there are no issues as the way it is, I'd rather not make any further changes.

mechPenSketch committed:
Merge branch 'master' into expose_hotzones

mechPenSketch commented:
Is merging conflict at this stage a mistake? Ever since the master branch merges into this one, I can't find my past commits in Github Desktop's history section. Being buried underneath make it harder to find and squash them.

pycbouh commented:
You shouldn't merge master in your branch, you need to rebase your branch against the master instead. You also need to squash your commits into one.

mechPenSketch commented:
Ok, so resolving conflict when prompt is a bad move. While I try to fix this, you've mentioned that my documentation need some amendment. I have edited it following my code fixes, but are there any more changes I need to make?

mechPenSketch committed:
Revert "Merge branch 'master' into expose_hotzones"

This reverts commit 8e6a9b8110953d9ad45ec6ba2a93f752f48eaa34, reversing
changes made to 8df9bf784c6e7d3227acd940688fb8fe8a289286.

mechPenSketch commented:
Closing as I've made a squashed PR.
